### PR TITLE
[OCRVS-10266] Implement  and  helpers for certificates, add print action optimistically into the array so its details can be read

### PR DESCRIPTION
**Pull request by rikukissa**
# New handlebar helpers

## $action
```
{{ $action "DECLARE" }}
```

Gives you the details of the **latest** action with type "DECLARE". Can be used in combination with `$lookup` to access data fields

```
{{ $lookup ($action "DECLARE") "createdAt" }}
```

## $actions
```
{{ $actions "DECLARE" }}
```

Gives you a **list** of actions with a certain type. Can be for instance used in if statements where the number of previous actions is needed

```
{{ $lookup ($actions "DECLARE") "length" }}
```

## $json

```
{{ $json ($action "DECLARE") }}
```

A new debugging helper for printing a result of another handlebar into the certificate. Useful when you just want to know what fields are available in the data.

## Checklist

- [ ] I have linked the correct Github issue under "Development"
- [ ] I have tested the changes locally, and written appropriate tests
- [ ] I have tested beyond the happy path (e.g. edge cases, failure paths)
- [ ] I have updated the changelog with this change (if applicable)
- [ ] I have updated the GitHub issue status accordingly

### CODE REVIEW

File: packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts [hunk 1]
```diff
@@ -14,7 +14,7 @@ export const tennisClubMembershipCertifiedCertificateTemplate = `
   <g clip-path="url(#clip0_5_2)">
     <path d="M583 0H0V842H583V0Z" fill="white" />
     <text fill="#222222" xml:space="preserve" style="white-space: pre" font-family="Noto Sans" font-size="8" letter-spacing="0em">
-      <tspan x="62" y="426.352">{{ $lookup $metadata "modifiedAt"}}</tspan>
+      <tspan x="62" y="426.352">{{ $lookup ($action "PRINT_CERTIFICATE") "createdAt" }}</tspan>
```
Comment (rikukissa) [inline packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts:17]: After this change, this will be the officially supported way to get the `Date of certification` to a certificate. FYI @tumbledwyer @Siyasanga

Comment (Zangetsu101) [reply packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts:17]: @tumbledwyer This should remove the need for `copiesPrintedForTemplate`

Comment (Zangetsu101) [reply packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts:17]: or not, as we can't filter on the action values yet

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.test.ts
Comment (Zangetsu101) [inline packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.test.ts:1]: Nice one 👍

File: packages/client/src/v2-events/hooks/useFormDataStringifier.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ import {
   isDateFieldType,
   isHttpFieldType
 } from '@opencrvs/commons/client'
+import { ActionDocument } from '@opencrvs/commons'
```
Comment (Zangetsu101) [inline packages/client/src/v2-events/hooks/useFormDataStringifier.ts]: is this being used somewhere?

File: packages/client/src/v2-events/hooks/usePrintableCertificate.ts [hunk 3]
```diff
@@ -80,15 +84,40 @@ export const usePrintableCertificate = ({
...
+    createdAt: new Date().toISOString(),
+    createdBy: userFromUsersList.id,
+    createdByRole: userFromUsersList.role,
+    status: 'Accepted',
+    declaration: {},
+    annotation: null,
+    originalActionId: null,
+    createdBySignature: userFromUsersList.signature,
+    createdAtLocation: userDetails.primaryOffice.id as UUID
+  } satisfies PrintCertificateAction)
```
Comment (Zangetsu101) [inline packages/client/src/v2-events/hooks/usePrintableCertificate.ts]: Shouldn't this also include the `content.templateId`

**Review by Zangetsu101 — APPROVED** [review]

File: packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts (hunk 1 above)
Comment (rikukissa) [reply packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts:17]: I recommend using a custom handlebar helper in country config for these sort of use cases so we can keep the core API as generic as we can

File: packages/client/src/v2-events/hooks/useFormDataStringifier.ts (hunk 2 above)
Comment (rikukissa) [reply packages/client/src/v2-events/hooks/useFormDataStringifier.ts]: Good catch!

File: packages/client/src/v2-events/hooks/usePrintableCertificate.ts (hunk 3 above)
Comment (rikukissa) [reply packages/client/src/v2-events/hooks/usePrintableCertificate.ts]: It indeed should!

File: packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts (hunk 1 above)
Comment (Zangetsu101) [reply packages/client/.storybook/tennisClubMembershipCertifiedCertificateTemplate.ts:17]: Agreed, if we expose the `content` property from the current handlebars, it should be fairly simple to create a custom handlebar in countryconfig for `copiesPrinted`

File: packages/client/src/v2-events/hooks/usePrintableCertificate.ts (hunk 3 above)
Comment (Zangetsu101) [reply packages/client/src/v2-events/hooks/usePrintableCertificate.ts]: Maybe it should have actually been a required property

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts [hunk 4]
```diff
@@ -201,10 +202,22 @@ function isMessageDescriptor(obj: unknown): obj is MessageDescriptor {
 
 const cache = createIntlCache()
 
+function isActionDocument(obj: unknown): obj is ActionDocument {
+  return (
+    typeof obj === 'object' &&
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts]: `ActionDocument.parse` or safeParse

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts [hunk 5]
```diff
@@ -247,6 +261,39 @@ export function compileSvg({
...
+   * Handlebars helper: $actions
+   *
+   * Resolves all actions for a specific action type
+   *
+   * @param actionType - The type of action to look up (e.g., "PRINT_CERTIFICATE")
+   * @returns The resolved list of actions
+   *
+   * @example {{ $actions "PRINT_CERTIFICATE" }}
+   */
+  function $actionsFn(actionType: string) {
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts:267]: What happens in the case where action is initially rejected and resubmitted? Should these be traversed from right / ordered newestFirst

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts [hunk 6]
```diff
@@ -247,6 +261,39 @@ export function compileSvg({
...
+   *
+   * Finds the latest action data for a specific action type and property path.
+   *
+   * @param actionType - The type of action to look up (e.g., "PRINT_CERTIFICATE")
+   * @returns The resolved value from the action data
+   *
+   * @example {{ $action "PRINT_CERTIFICATE" }}
+   */
+
+  function $action(actionType: string) {
```
Comment (makelicious) [inline packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts:286]: here we are, very nice. Does this need to take accepted| rejected | requested into account?

Could you review jamil's work here that changes it even further.

https://github.com/opencrvs/opencrvs-core/pull/10246

**Review by makelicious — COMMENTED** [review]
good stuff. Check the "duplicate action" things on jamilur's pr if it has any effect.

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts (hunk 5 above)
Comment (rikukissa) [reply packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts:267]: This would be the one that returns all actions, so result could have multiple instances of declare for instance

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts (hunk 6 above)
Comment (rikukissa) [reply packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts:286]: I changed this now so it only accepts Accepted ones. No need or the client even to know about the other statuses

File: packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts [hunk 7]
```diff
@@ -272,11 +316,52 @@ export function compileSvg({
...
+          locations,
+          action.data.createdAtLocation
+        ),
+        createdByRole: action.data.createdByRole
+      }
+
+      return getMixedPath(resolvedAction, propertyPath)
+    }
+
+    return obj[propertyPath as keyof typeof obj] ?? ''
```
Comment (Zangetsu101) [inline packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts]: Will `$lookup` work with arrays? If so maybe we should put an example of how the path would look like

Comment (rikukissa) [reply packages/client/src/v2-events/features/events/actions/print-certificate/pdfUtils.ts]: Not by default, although of course you could do `{{ $lookup ['a'] 0 }}` but I think by default we would only use it for objects
